docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+  Protocol api.Protocol `json:"protocol"`
+
+  // If specified, the port on the given protocol.  This can 
+  // either be a numerical or named port.  If not defined,
+  // this NetworkPolicyPort does not restrict based on port number.
+  // If defined, only traffic on the specified protocol AND port
+  // will be matched by this NetworkPolicyPort.
+  Port IntOrString `json:"port,omitempty"`
+}
+
+type NetworkPolicySource struct {
+  // If 'Namespaces' is defined, 'Pods' must not be.
+  // This is a label selector which selects Pods in this namespace.
+  // This NetworkPolicySource matches any pods selected by this selector.
+  Pods *unversioned.LabelSelector `json:"pods,omitempty"`
+
+  // If 'Pods' is defined, 'Namespces' must not be.
+  // Selects Kubernetes Namespaces.  This NetworkPolicySource matches 
+  // all pods in all namespaces selected by this label selector. 
+  Namespaces *unversioned.LabelSelector `json:"namespaces,omitempty"`
+}
+```
+
+### Behavior
+
+The following pseudo-code attempts to define when traffic is allowed to a given pod when using this API.
+
+```python
+def is_traffic_allowed(traffic, pod):
+  """
+  Returns True if traffic is allowed to this pod, False otherwise.
+  """
+  if namespace.network_isolation == False:
+    # If network-isolation is disabled, all traffic is allowed.
+    return True 
+  elif traffic.source == pod.node:
+    # Traffic is from the pod's host - this allows for kubelet health checks.
+    return True
+  else:
+    # If network-isolation is enabled, only allow traffic 
+    # that matches a network policy which selects this pod.
+    for network_policy in all_network_policies:
+      if not network_policy.podSelector.selects(pod):
+        # This policy doesn't select this pod. Try the next one. 
+        continue
+
+      # This policy selects this pod.  Check each ingress rule 
+      # defined on this policy to see if it allows the traffic.
+      for ingress_rule in network_policy.ingress:
+        if ingress_rule.from.matches(traffic) and \
+             ingress_rule.ports.matches(traffic):
+          return True 
+
+  # network-isolation is True and no policies match the given pod and traffic.
+  return False
+```
+
+### Open Questions
+
+> The question then remains whether one NP object per "app" makes sense (in this case app={a,b,c}) or whether one NP object per selector
+
+- A single podSelector per NetworkPolicy may lead to managing a large number of NetworkPolicy objects, each of which is small and easy to understand on its own.  Allowing a podSelecor per ingress rule would allow for fewer, larger NetworkPolicy objects that may be easier to manage, but harder to understand individually.  This proposal has opted to favor a larger number of smaller objects that are easier to understand.  Is this the right decision? 
+
+- Is the `Namespaces` selector in the `NetworkPolicySource` struct too coarse? Do we need to support the AND combination of `Namespaces` and `Pods`? 
```

Comment:
I am not familiar with how the world uses Namespaces.  Does it even use them at all?

Unless presented with strong evidence of a need to do otherwise, my inclination is to not let one namespace "see" into the details of another.

BTW, I did notice any explicit statement about what it means if a `NetworkPolicySource` has _neither_ `Pods` nor `Namespaces`.  Since the thinking here is intersection, I suppose the case of zero fields would have to mean the `NetworkPolicySource` matches all sources.  Note that we already have another way of expressing that.  Do we want to allow both ways?  If not, which is better?  In my opinion, an empty `NetworkPolicySource` is better, because the other involves an irregularity (a missing `NetworkPolicyIngressRule.From` matches everything but the natural default for an additive list is to match nothing; while it is true that we have this irregularity in other places too, each one is a cognitive speed bump).